src/main/java/bio/terra/stairway/FlightMap.java

```diff
@@ -77,33 +102,78 @@ public void makeImmutable() {
    * @throws ClassCastException if found, not castable to the requested type
    */
   public <T> T get(String key, Class<T> type) {
-    Object value = map.get(key);
+    String value = map.get(key);
     if (value == null) {
       return null;
     }
 
-    if (type.isInstance(value)) {
-      return type.cast(value);
+    try {
+      return getObjectMapper().readValue(value, type);
+    } catch (final JsonProcessingException ex) {
+      throw new ClassCastException(
+          "Found value '" + value + "' is not an instance of type " + type.getName());
     }
-    throw new ClassCastException(
-        "Found value '" + value.toString() + "' is not an instance of type " + type.getName());
   }
 
+  /**
+   * Gets the raw (serialized) data stored for an Object in the hash map.
+   *
+   * @param key Key to lookup the data for in the map.
+   * @return String mapped to key, or null if key does not exist.
+   */
+  public String getRaw(String key) {
+    return map.get(key);
+  }
+
+  /**
+   * Serialize an Object and place it into the map at a given key.
+   *
+   * @param key Key to store the object under.
+   * @param value Object to serialize and store.
+   * @throws JsonConversionException if object could not be serialized.
+   */
   public void put(String key, Object value) {
+    try {
+      map.put(key, getObjectMapper().writeValueAsString(value));
+    } catch (final JsonProcessingException ex) {
+      throw new JsonConversionException("Failed to convert object to json string", ex);
+    }
+  }
+
+  /**
+   * Stores a raw (serialized) data value for a given key.
+   *
+   * @param key Key to store the data under.
+   * @param value Data to store.
+   */
+  public void putRaw(String key, String value) {
     map.put(key, value);
   }
 
   public String toJson() {
     try {
-      return getObjectMapper().writeValueAsString(map);
-    } catch (JsonProcessingException ex) {
+      // Serializing as Map<String, RawJsonValue> will produce a format compatible with older code's
+      // Map<String, Object> representation.
+      Map<String, RawJsonValue> temporaryMap = new HashMap<>();
+      for (Map.Entry<String, String> entry : map.entrySet()) {
+        temporaryMap.put(entry.getKey(), RawJsonValue.create(entry.getValue()));
+      }
+      return getObjectMapper().writeValueAsString(temporaryMap);
+    } catch (final JsonProcessingException ex) {
       throw new JsonConversionException("Failed to convert map to json string", ex);
     }
   }
 
   public void fromJson(String json) {
     try {
-      map = getObjectMapper().readValue(json, new TypeReference<Map<String, Object>>() {});
+      // Deserialize Map<String, Object> written by previous versions of code; deserializing as
+      // Map<String, JsonNode> allows us to defer deserialization and obtain String representations
+      // to store in our Map<String, String>.
```

Comment:
I'm curious about the interaction of this readValue with the embedded typing in the StairwayMapper.
Do we need to use a mapper that will not process the embedded types?

Reply:
What this does is deserializes each value in the map to the root node of a Jackson tree, which is either a string/int for primitives ("value node") or an array ("array node") for types (entry 0 is a string representing the type, entry 1 is the payload as a JSON object.

Calling `toString()` on the root node renders the JSON, which is what we store in our `Map<String, String>`.

This is the closest thing to official docs I could find on Jackson tree model: https://www.baeldung.com/jackson-json-node-tree-model

Reply:
Reverted per previous comment.